sql-plugin/src/main/scala/com/nvidia/spark/rapids/spill/SpillFramework.scala

```diff
@@ -1597,59 +1590,71 @@ object SpillFramework extends Logging {
 /**
  * A bounce buffer wrapper class that supports the concept of acquisition.
  *
- * The bounce buffer is acquired exclusively. So any calls to acquire while the
- * buffer is in use will block at `acquire`. Calls to `release` notify the blocked
- * threads, and they will check to see if they can acquire.
+ * The bounce buffer is acquired from a BounceBufferPool, so any calls to
+ * BounceBufferPool.nextBuffer will block if the pool has no available buffers.
  *
- * `close` is the interface to unacquire the bounce buffer.
+ * `close` restores a bounce buffer to the pool for other callers to use.
  *
- * `release` actually closes the underlying DeviceMemoryBuffer, and should be called
+ * `release` actually closes the underlying buffer, and should be called
  * once at the end of the lifetime of the executor.
  *
- * @param dmb - actual cudf DeviceMemoryBuffer that this class is protecting.
+ * @param buf - actual cudf DeviceMemoryBuffer that this class is protecting.
+ * @param pool - the pool to which this buffer belongs
  */
-private[spill] case class DeviceBounceBuffer(var dmb: DeviceMemoryBuffer) extends AutoCloseable {
-  private var acquired: Boolean = false
-  def acquire(): DeviceBounceBuffer = synchronized {
-    while (acquired) {
-      wait()
-    }
-    acquired = true
-    this
-  }
-
-  private def unaquire(): Unit = synchronized {
-    acquired = false
-    notifyAll()
-  }
-
+private[spill] class BounceBuffer[T <: AutoCloseable](
+                                                       var buf: Option[T],
```

Comment:
nit: What reason is there to have `buf` be optional? I get we want to remove it after we have released it, but we have use after free checking in DeviceMemoryBuffer and the rest already. we could also set it to null. I just find it a little cleaner to not have it be an `Option`.

Reply:
The compiler didn't like me setting the generic type `T` to `null` because of a type mismatch, but I can get it to work with `null.asInstanceOf[T]`. I agree it doesn't need to be an option, I can change it.